### ISSUE TITLE
Benchmark vs Image vs FastImage

### ISSUE THREAD

**Issue by henrymoulton** [mralj/react-native-mad-image#1]
Hi I saw your insightful comment here: https://github.com/DylanVann/react-native-fast-image/issues/715#issuecomment-674124936

and I wondered if you'd created `react-native-mad-image` to solve some of the issues with FastImage. Just wondering if why you created it and if it offers any performance advantages.

I also noticed you said this interesting line here:
> We have noticed in our app that for displaying local images, react native's default Image actually performs better :)

I wondered if you'd been able to figure out why this is?

**Comment by mralj** [comment]
Hey,
Sorry for not replying sooner, I completely missed this notification :/

Regarding the `react-native-mad-image` it was my idea to solve issues with `fast-image`, unfortunately never found time to do this :/
I wanted to create wrappers around Nuke (for iOS) and Glide (for Android).

We are still using `react-native-fast-image` in our app, but we have our own fork (which isn't too different from main repo): https://github.com/Madbarz/react-native-fast-image

The only advantage you get with our fork (although, IMO it's a big one) is much better perf. on Android, because images are immediately resized to save memory and since they have smaller dimensions it's easier to render them.

This lead to issue on some androids, having poorer image quality so we had workout around that as well 😅
So here is example how we use our fork:

``` TS
    const androidShouldScaleBasedOnDPI = Platform.OS === "android" && resizeImageBasedOnDPIOnAndroid;
    const imageHeight = androidShouldScaleBasedOnDPI ? scaleValueBasedOnDPI(style.height) : style.height;
    const imageWidth = androidShouldScaleBasedOnDPI ? scaleValueBasedOnDPI(style.width) : style.width;
    const resizeImageAndroid = Platform.OS === "android" ? { width: imageWidth, height: imageHeight } : undefined;

          <FastImage
            source={calculatedSource}
            style={style}
            resizeMode={resizeMode as ResizeMode}
            resizeImageAndroid={resizeImageAndroid}
            testID={testID}
          />
```

Basically we have _facade_ component around Image/FastImage and we can choose what we want to render, also we pass this  prop `resizeImageBasedOnDPIOnAndroid` if the image quality on android is not good.

If you want, I can paste here whole code for Image component we use in our app :)


----------


>>We have noticed in our app that for displaying local images, react native's default Image actually performs better :)

>I wondered if you'd been able to figure out why this is?

No, unfortunately not, but we didn't dig deeper into this :)

**Comment by romantseg** [comment]
@mralj thank you for this! This has really helped us debugging Android performance. Would you mind pasting the whole code for Image component?

**Comment by mralj** [comment]
Sure!

Here is the code for the component:

```TSX
import React from "react";
import { Image, Platform } from "react-native";
import FastImage, { ResizeMode } from "react-native-fast-image";
import { ImageType, ImageUniversalType } from "src/components/images/ImageUniversalModel";
import { calculateImageSource } from "src/components/images/ImageUniversalService";

const ImageUniversal = ({
  source,
  style,
  type = ImageType.FastImage,
  resizeMode = FastImage.resizeMode.cover,
  testID = "plain_image",
  resizeImageBasedOnDPIOnAndroid
}: ImageUniversalType) => {
  const shouldForceFastImageForAndroid = Platform.OS === "android";
  const imageType =
    shouldForceFastImageForAndroid || type === ImageType.FastImage ? ImageType.FastImage : ImageType.Image;

  const { imageSource, fastImageSource, resizeImageAndroid } = calculateImageSource({
    source,
    style,
    resizeImageBasedOnDPIOnAndroid
  });

  switch (imageType) {
    case ImageType.Image:
      return <Image source={imageSource} style={style} resizeMode={resizeMode} testID={`${testID}_id`} />;
    case ImageType.FastImage:
      return (
        <FastImage
          source={fastImageSource}
          style={style}
          resizeMode={resizeMode as ResizeMode}
          resizeImageAndroid={resizeImageAndroid}
          testID={testID}
        />
      );
  }
};

function reRenderImageOnlyIfSourceChanges(prevProps: ImageUniversalType, nextProps: ImageUniversalType) {
  return prevProps.source === nextProps.source;
}

export default React.memo(ImageUniversal, reRenderImageOnlyIfSourceChanges);
```

And here is the code for the service (which has defined `calculateImageSource`):

```TSX
import { Platform } from "react-native";
import { CalculatedSource, CalculateImageSource } from "src/components/images/ImageUniversalModel";
import {
  ANDROID_SCREEN_DENSITY_INFO_BASED_ON_PIXEL_RATIO,
  IMAGE_SCALE_BASED_ON_PIXEL_RATIO
} from "src/utils/responsiveScreen";

export function calculateImageSource({
  resizeImageBasedOnDPIOnAndroid,
  style,
  source
}: CalculateImageSource): CalculatedSource {
  const androidShouldScaleBasedOnDPI = Platform.OS === "android" && resizeImageBasedOnDPIOnAndroid;

  const imageHeight = androidShouldScaleBasedOnDPI
    ? scaleValueBasedOnDPI(style.height, ANDROID_SCREEN_DENSITY_INFO_BASED_ON_PIXEL_RATIO.height)
    : style.height;
  const imageWidth = androidShouldScaleBasedOnDPI
    ? scaleValueBasedOnDPI(style.width, ANDROID_SCREEN_DENSITY_INFO_BASED_ON_PIXEL_RATIO.width)
    : style.width;

  const uriIsPassedIn = typeof source === "string";
  const calculatedSource = uriIsPassedIn ? { uri: source, height: imageHeight, width: imageWidth } : source;
  const calculatedFastImageSource = uriIsPassedIn ? { uri: source } : source;
  const resizeImageAndroid = Platform.OS === "android" ? { width: imageWidth, height: imageHeight } : undefined;

  return { imageSource: calculatedSource, fastImageSource: calculatedFastImageSource, resizeImageAndroid };
}

function scaleValueBasedOnDPI(value: number, recommendedValueForAndroidDensity: number) {
  const valueScaled = value * IMAGE_SCALE_BASED_ON_PIXEL_RATIO;

  return valueScaled > recommendedValueForAndroidDensity ? recommendedValueForAndroidDensity : valueScaled;
}
```

Please note we are using our fork instead of the _main repo_.  
And if you need code for the
 `ANDROID_SCREEN_DENSITY_INFO_BASED_ON_PIXEL_RATIO, IMAGE_SCALE_BASED_ON_PIXEL_RATIO` which we have defined in our `responsiveScreen` do tell 🙂 (the reason I'm not pasting it immediately is, it has a bit more code, and not sure will it be of relevance/interest to you)

**Comment by romantseg** [comment]
@mralj thank you for sharing this and answering so fast! Appreciate both the code and the style of implementation, it's a fascinating example. If possible, would you mind sharing code for `ANDROID_SCREEN_DENSITY_INFO_BASED_ON_PIXEL_RATIO`, `IMAGE_SCALE_BASED_ON_PIXEL_RATIO`

**Comment by mralj** [comment]
I'm glad you liked the code 😁  Now for the `responsiveScreen` we have 2 files:

The model:

```TSX
export type ScreenSizeValue = number;
export type ScreenSizeValueExtended = number | string | boolean;

export type ScreenSizesHeight = {
  name: "m" | "l" | "xl" | "xxl";
  height: number;
};

export type ScreenWidthSizes = {
  name: "xs" | "s" | "m" | "l" | "xl" | "xxl" | "xxxl";
  width: number;
};

export type ScreenAspectRatioSize = Pick<ScreenWidthSizes, "name"> & { ratio: number; orderForSorting: number };

export type ScreenSizeOptions = {
  // iPhone 6
  m?: ScreenSizeValue;
  // iPhone 7 Plus
  l?: ScreenSizeValue;
  // iPad Air 2/Mini 4
  xl?: ScreenSizeValue;
  // iPad Pro (10.5 inch)
  xxl?: ScreenSizeValue;
};

export type ExtendedScreenSizeOptions<T extends ScreenSizeValueExtended> = {
  xs: T;
  s?: T;
  m?: T;
  l?: T;
  xl?: T;
  xxl?: T;
  xxxl?: T;
};

export type ScreenSizeOptionsPerPlatform<T extends ScreenSizeValueExtended> = {
  ios: ExtendedScreenSizeOptions<T>;
  android: ExtendedScreenSizeOptions<T>;
};

// Height in Points
export const screenHeightSizes: ScreenSizesHeight[] = [
  {
    name: "xxl",
    height: 1365
  },
  {
    name: "xl",
    height: 1023
  },
  {
    name: "l",
    height: 730
  },
  {
    name: "m",
    height: 666
  }
];

export const screenAspectRatioSizes: ScreenAspectRatioSize[] = [
  {
    name: "xl",
    ratio: 1.4,
    orderForSorting: 1000
  },
  {
    name: "l",
    ratio: 1.7,
    orderForSorting: 2000
  },
  {
    name: "xs",
    ratio: 1.901,
    orderForSorting: 5000
  },
  {
    name: "s",
    ratio: 2.1,
    orderForSorting: 4000
  },
  {
    name: "m",
    ratio: 3,
    orderForSorting: 3000
  }
];

export const AndroidScreenDensity = {
  ldpi: {
    density: 0.75,
    width: 420,
    height: 320
  },
  mdpi: {
    density: 1,
    width: 320,
    height: 480
  },
  hdpi: {
    density: 1.5,
    width: 480,
    height: 800
  },
  xhdpi: {
    density: 2,
    width: 640,
    height: 960
  },
  xxhdpi: {
    density: 3,
    width: 960,
    height: 1600
  },
  xxxhdpi: {
    density: 4,
    width: 1280,
    height: 1920
  }
};

```

And here is the "main thing"

```TSX
import { Dimensions, PixelRatio, Platform } from "react-native";
import DeviceInfo from "react-native-device-info";
import { isDeviceTablet } from "src/utils/deviceInfo";
import { sanitizeNumericInput } from "src/utils/helpers";
import {
  AndroidScreenDensity,
  ExtendedScreenSizeOptions,
  ScreenAspectRatioSize,
  screenAspectRatioSizes,
  screenHeightSizes,
  ScreenSizeOptions,
  ScreenSizeOptionsPerPlatform,
  ScreenSizeValue,
  ScreenSizeValueExtended
} from "src/utils/responsiveScreenModel";

export const DEVICE_PIXEL_RATIO = PixelRatio.get();

export const SCREEN_WIDTH = Dimensions.get("screen").width;
export const SCREEN_HEIGHT = Dimensions.get("screen").height;
const WINDOW_HEIGHT = Dimensions.get("window").height;

export const BOTTOM_NAVIGATION_BAR_HEIGHT = SCREEN_HEIGHT - WINDOW_HEIGHT;

const isIPhoneWithNotch = () => {
  if (Platform.OS !== "ios" || Platform.isPad || Platform.isTVOS) {
    return false;
  }

  return DeviceInfo.hasNotch();
};

const isDeviceIPadWithNewIOS = () =>
  Platform.OS === "ios" && sanitizeNumericInput(Platform.Version) > 11 && isDeviceTablet;

const calculateBottomTabsSafeAreaHeight = () => {
  if (isIPhoneWithNotch()) {
    return 34;
  }

  return 0;
};

const calculateTopBarHeight = (_retrievedTabHeight = 0) => {
  let topBarHeight = 0;

  if (Platform.OS === "android") {
    topBarHeight = 56;
  } else if (isDeviceIPadWithNewIOS()) {
    topBarHeight = 50;
  } else {
    topBarHeight = 44;
  }

  return topBarHeight;
};

const calculateBottomTabHeight = (_retrievedTabHeight = 0, _bottomTabsSafeAreaHeight = 0) => {
  let retrievedTabHeight = 0;

  if (Platform.OS === "android") {
    retrievedTabHeight = 56;
  } else if (isDeviceIPadWithNewIOS()) {
    retrievedTabHeight = 50;
  } else if (isDeviceTablet) {
    retrievedTabHeight = 44;
  } else {
    retrievedTabHeight = 49;
  }

  return retrievedTabHeight;
};

const calculateStatusBarHeight = (_retrievedTabHeight = 0) => {
  let statusBarHeight = 0;

  if (Platform.OS === "android") {
    statusBarHeight = 24;
  } else if (isIPhoneWithNotch()) {
    statusBarHeight = 44;
  } else {
    statusBarHeight = 20;
  }

  return statusBarHeight;
};

// special case, doesn't depend on value from Navigation.constants
export const BOTTOM_TABS_SAFE_AREA_HEIGHT = calculateBottomTabsSafeAreaHeight();

export const TOPBAR_HEIGHT = calculateTopBarHeight(0);
export const BOTTOM_TABS_HEIGHT = calculateBottomTabHeight(0, BOTTOM_TABS_SAFE_AREA_HEIGHT);
export const STATUSBAR_HEIGHT = calculateStatusBarHeight(0);

const MATCHED_SCREEN_HEIGHTS = screenHeightSizes.filter(screenSize => {
  return SCREEN_HEIGHT >= screenSize.height;
});

// Example usage:
// getSizeFromConfig({ l: 12, m: 8}, 16)
// getSizeFromConfig({xl: 12}, 16)
function getSizeFromConfig(screenSizeOptions: ScreenSizeOptions, defaultValue: ScreenSizeValue): ScreenSizeValue {
  const matchedSize = MATCHED_SCREEN_HEIGHTS.find(matchedScreenSize =>
    screenSizeOptions.hasOwnProperty(matchedScreenSize.name)
  );

  if (matchedSize) {
    return screenSizeOptions[matchedSize.name] as ScreenSizeValue;
  }

  return defaultValue;
}

const foundAspectRatio = screenAspectRatioSizes.find(({ ratio }) => {
  return SCREEN_HEIGHT / SCREEN_WIDTH < ratio;
}) as ScreenAspectRatioSize;

const MATCHED_ASPECT_RATIOS = screenAspectRatioSizes
  .filter(removeAllScreenSizesGreaterThanTheFound)
  .sort(sortRemainingScreenSizesIndirectlyByName);

function getResponsiveValuePerAspectRatio<T extends ScreenSizeValueExtended>(
  screenAspectRatio: ExtendedScreenSizeOptions<T>
): T {
  const matchedSize = MATCHED_ASPECT_RATIOS.find(matchedAspectRatio =>
    screenAspectRatio.hasOwnProperty(matchedAspectRatio.name)
  );

  if (matchedSize) {
    return screenAspectRatio[matchedSize.name] as T;
  }

  return screenAspectRatio.xs;
}

function getResponsiveValuePerAspectRatioPerPlatform<T extends ScreenSizeValueExtended>(
  screenSizeOptions: ScreenSizeOptionsPerPlatform<T>
): T {
  return Platform.OS === "ios"
    ? getResponsiveValuePerAspectRatio(screenSizeOptions.ios)
    : getResponsiveValuePerAspectRatio(screenSizeOptions.android);
}

function removeAllScreenSizesGreaterThanTheFound({ orderForSorting: order }: ScreenAspectRatioSize) {
  return foundAspectRatio.orderForSorting <= order;
}

function sortRemainingScreenSizesIndirectlyByName(
  { orderForSorting: firstOrder }: ScreenAspectRatioSize,
  { orderForSorting: secondOrder }: ScreenAspectRatioSize
) {
  return firstOrder - secondOrder;
}

function getAndroidScreenDensityInfoBasedOnPixelRatio() {
  const { xxxhdpi, xxhdpi, xhdpi, hdpi, mdpi, ldpi } = AndroidScreenDensity;

  if (DEVICE_PIXEL_RATIO >= xxxhdpi.density) {
    return xxxhdpi;
  }

  if (DEVICE_PIXEL_RATIO >= xxhdpi.density) {
    return xxhdpi;
  }

  if (DEVICE_PIXEL_RATIO >= xhdpi.density) {
    return xhdpi;
  }

  if (DEVICE_PIXEL_RATIO >= hdpi.density) {
    return hdpi;
  }

  if (DEVICE_PIXEL_RATIO >= mdpi.density) {
    return mdpi;
  }

  return ldpi;
}

function getImageScaleBasedOnPixelRatio() {
  const { xhdpi } = AndroidScreenDensity;

  return DEVICE_PIXEL_RATIO > xhdpi.density ? xhdpi.density : DEVICE_PIXEL_RATIO;
}

const ANDROID_SCREEN_DENSITY_INFO_BASED_ON_PIXEL_RATIO = getAndroidScreenDensityInfoBasedOnPixelRatio();
const IMAGE_SCALE_BASED_ON_PIXEL_RATIO = getImageScaleBasedOnPixelRatio();

export {
  getSizeFromConfig,
  isIPhoneWithNotch,
  getResponsiveValuePerAspectRatio,
  getResponsiveValuePerAspectRatioPerPlatform,
  ANDROID_SCREEN_DENSITY_INFO_BASED_ON_PIXEL_RATIO,
  IMAGE_SCALE_BASED_ON_PIXEL_RATIO
};

```

The main purpose of this `responsiveScreen` is not the `ImageUniversal` component, but we mainly use it for styling.
It's really neat for fine-tuning the styling for the different devices, example:

```TS
const cellExerciseTagText: TextStyle = {
  fontSize: SCREEN_WIDTH / 35,
  marginTop: getResponsiveValuePerAspectRatio({
    xs: SCREEN_HEIGHT / 200,
    s: SCREEN_HEIGHT / 350,
    l: SCREEN_HEIGHT / 100
  }),
  color: "#222222",
  opacity: 0.8
};
```

And another one:

```TS
  secondaryItemSubtitle: {
    color: "#959595",
    fontSize: SCREEN_WIDTH / 30,
    marginTop: getResponsiveValuePerAspectRatioPerPlatform({
      ios: { xs: SCREEN_HEIGHT / 300 },
      android: { xs: 0, m: SCREEN_HEIGHT / 500 }
    }),
    padding: 0,
    maxWidth: SCREEN_WIDTH / 1.4
  }
```
